### PR TITLE
Add a command line tool for load testing the Provisioner

### PR DESCRIPTION
**Pull request by gigawhitlocks**
#### Summary

This PR introduces a new tool in the `cmd/tools` directory which can be installed with `go install ./cmd/tools/cloudburst` from the project base directory.

`cloudburst` is a command line tool for load testing the Provisioner. `cloudburst` allows you to very quickly spin up a huge number of Installations and collects timing data as it runs.

```
Run a load test against a provisioning server.

Usage:
  cloudburst [flags]

Flags:
      --batch int       Specify the number of Installations in each batch. Installations in a batch are install serially and batches are installed in parallel. (default 5)
  -h, --help            help for cloudburst
      --runs int        Number of times to repeat the test (default 1)
      --server string   Location of the Provisioning Server to load test (default "http://localhost:8075")
      --total int       Number of Installations to provision (default 20)
```

The nature of this PR is such that this code won't get "stale" the way a more tightly integrated PR might, so feel free to make larger requests in this PR than normal, as my primary goal here is to provide a tool that has an immediate as well as long term value payoff.

#### Ticket Link

No ticket; addresses "explosive scaling testing" OKR goal.

#### Release Note

```release-note
Added cloudburst, a tool for explosive scaling / load testing.
```

### CODE REVIEW

File: cmd/tools/cloudburst/cloudburst.go [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+		err := blaster.createGroup()
+		if err != nil {
+			return errors.Wrap(err, "failed to create group for test installations")
+		}
+
+		logger.Infof("Waiting for Group to be created..")
+		blaster.waitForGroup()
+
+		allReports := []Report{}
+		for i := 0; i < runs; i++ {
```
Comment (gigawhitlocks) [inline cmd/tools/cloudburst/cloudburst.go:103]: Forgot to refactor this bit before opening the PR. Will do that.

File: cmd/tools/cloudburst/cloudburst.go [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+			logger.WithError(err).Errorf("failed to get group %s", b.group.ID)
+			continue
+		}
+		if g == nil {
+			continue
+		}
+		return
+	}
+}
+func init() {
```
Comment (gabrieljackson) [inline cmd/tools/cloudburst/cloudburst.go:268]: nit: `init()` near the top of the file would be nice.

File: cmd/tools/cloudburst/cloudburst.go [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+	logger = log.New()
+	blastCommand.PersistentFlags().String("server", "http://localhost:8075", "Location of the Provisioning Server to load test")
+	blastCommand.PersistentFlags().Int("runs", 1, "Number of times to repeat the test")
+	blastCommand.PersistentFlags().Int("batch", 5, "Specify the number of Installations in each batch. Installations in a batch are install serially and batches are installed in parallel.")
```
Comment (gabrieljackson) [inline cmd/tools/cloudburst/cloudburst.go:272]:
```suggestion
	blastCommand.PersistentFlags().Int("batch", 5, "Specify the number of Installations in each batch. Installations in a batch are installed serially and batches are installed in parallel.")
```

File: cmd/tools/cloudburst/cloudburst.go [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+	}
+
+	return allInstallations
+}
+
+// createInstallation requests creation of a single installation and
+// returns an error if the API does. It does not block or retry.
+func (b *Blaster) createInstallation() (*cloud.Installation, error) {
+	installationDTO, err := b.client.CreateInstallation(
+		&cloud.CreateInstallationRequest{
```
Comment (gabrieljackson) [inline cmd/tools/cloudburst/cloudburst.go:383]: Being able to specify `version`, `filestore`, and `database` with flags in the future would be nice.

**Review by gabrieljackson — APPROVED** [review]
Very cool. This will definitely be helpful to have around.

File: cmd/tools/cloudburst/cloudburst.go [hunk 5]
```diff
@@ -0,0 +1,414 @@
...
+		logger.Infof("Server address %s", serverAddress)
+		logger.WithField("server", serverAddress)
+		blaster := NewBlaster(serverAddress)
+		err := blaster.createGroup()
+		if err != nil {
+			return errors.Wrap(err, "failed to create group for test installations")
+		}
+
+		logger.Infof("Waiting for Group to be created..")
+		blaster.waitForGroup()
```
Comment (Szymongib) [inline cmd/tools/cloudburst/cloudburst.go:100]: Isn't the group creation synchronous operation? I am not sure if we need to wait for a group to be created.

**Review by Szymongib — APPROVED** [review]

File: cmd/tools/cloudburst/cloudburst.go (hunk 5 above)
Comment (gigawhitlocks) [reply cmd/tools/cloudburst/cloudburst.go:100]: It is not synchronous however the aggressive retrial strategy in the routines that send `CreateInstallationRequest`s does probably make this redundant. I will double-check and remove this if I find that it is not necessary.

File: cmd/tools/cloudburst/cloudburst.go (hunk 4 above)
Comment (gigawhitlocks) [reply cmd/tools/cloudburst/cloudburst.go:383]: https://mattermost.atlassian.net/browse/MM-30529 Created this story to track this work in the future

**Comment by gigawhitlocks** [comment]
Sorry about the wrecked history on this PR. I moved `init()` to satisfy one of Gabe's comments and rebased against `master` instead of merging, without thinking. Bleh.

**Pull request merged**